File: Codes/005_longestPalindrome/5.最长回文子串.go
/*
 * @Author: Theo_hui
 * @Date: 2021-08-27 16:25:31
 * @Descripttion:
 */
/*
 * @lc app=leetcode.cn id=5 lang=golang
 *
 * [5] 最长回文子串
 */

// @lc code=start

func expend(s string, left, right int) (int, int) {
	for {
		if left >= 0 && right < len(s) && s[left] == s[right] {
			left--
			right++
		} else {
			break
		}
	}
	return left + 1, right - 1
}

func longestPalindrome(s string) string {
	//中心拓展法

	//记录最值情况
	best_r, best_l := 0, 0

	//遍历
	for i, _ := range s {
		l1, r1 := expend(s, i, i)
		l2, r2 := expend(s, i, i+1)

		var r, l int

		if (r1 - l1) > (r2 - l2) {
			r, l = r1, l1
		} else {
			r, l = r2, l2
		}

		if (r - l) > (best_r - best_l) {
			best_l, best_r = l, r
		}
	}

	return s[best_l : best_r+1]

}

// @lc code=end

File: Codes/005_longestPalindrome/5.最长回文子串.js
/*
 * @Author: Theo_hui
 * @Date: 2021-08-27 16:44:58
 * @Descripttion: 
 */
/*
 * @lc app=leetcode.cn id=5 lang=javascript
 *
 * [5] 最长回文子串
 */

// @lc code=start
/**
 * @param {string} s
 * @return {string}
 */
var longestPalindrome = function(s) {

    // 子函数
    var expend = function(s,left,right){
        while(left>=0 && right<s.length && s[left]==s[right]){
            left--
            right++
        }
        return [left+1,right-1]
    }

    //记录最佳
    best_l=best_r=0

    for(var i=0;i<s.length;i++){
        let [l1,r1] = expend(s,i,i)
        let [l2,r2] = expend(s,i,i+1)
        let [r,l] = (r1-l1)>(r2-l2)?[r1,l1]:[r2,l2]
        if((r-l)>(best_r-best_l)){
            best_l=l
            best_r=r
        }
    }

    return s.substring(best_l,best_r+1)
};
// @lc code=end

